Maple2.File.IO/Nif/NifDocument.cs

```diff
@@ -0,0 +1,247 @@
+    }
+
+    public T? ReadBlockRef<T>() {
+        int index = Reader.ReadAdjustedInt32();
+
+        if (index != -1) {
+            return GetBlock<T>(index);
+        }
+
+        return default;
+    }
+
+    public List<T> ReadBlockRefList<T>() {
+        List<T> blocks = new List<T>();
+
+        int count = Reader.ReadAdjustedInt32();
+
+        blocks.EnsureCapacity(count);
+
+        for (int i = 0; i < count; ++i) {
+            T? block = ReadBlockRef<T>();
+
+            if (block is null) {
+                throw new InvalidCastException($"Null/invalid block handle in block list <{typeof(T).Name}> of size {count}");
+            }
+
+            blocks.Add(block);
+        }
+
+        return blocks;
+    }
+
+    public string ReadString() {
+        uint stringIndex = Reader.ReadAdjustedUInt32();
+
+        return header.Strings[stringIndex];
+    }
+
+    public void Reading(NifBlock block) {
+        ReadingBlock = block;
+    }
+
+    public void Parse() {
+        try {
+            ParseImplementation();
+        } catch (Exception ex) {
+            StringBuilder builder = new StringBuilder();
+
+            builder.Append($"Error reading nif: [{RelPath}]");
+
+            if (VersionString.Length > 0) {
+                builder.Append($" ({VersionString})");
+            }
+
+            if (ReadingBlock is not null) {
+                builder.AppendLine($" in block [{ReadingBlock.BlockIndex}] {ReadingBlock.BlockType} \"{ReadingBlock.Name}\"");
+            }
+
+            throw new Exception(builder.ToString(), ex);
+        }
+    }
+
+    private void ParseImplementation() {
+        int index = ReadHeaderString();
+
+        if (header.Version[0] < 30) {
+            throw new NifVersionNotSupportedException($"[{RelPath}]: NIF version number too low: {header.HeaderString}. Parser is built for v30+");
+        }
+
+        index += 4;
+
+        bool isLittleEndian = fileData[index] != 0;
+
+        MemoryStream stream = new MemoryStream(fileData);
+        Reader = new EndianReader(stream, isLittleEndian != BitConverter.IsLittleEndian);
+
+        Reader.Advance(index + 1);
+
+        uint userVersion = Reader.ReadAdjustedUInt32();
+        int numBlocks = Reader.ReadAdjustedInt32();
+        int metaBlockSize = Reader.ReadAdjustedInt32();
+
+        Reader.Advance(metaBlockSize);
+
+        ushort numBlockTypes = Reader.ReadAdjustedUInt16();
+
+        if (numBlockTypes == 0) {
+            return;
+        }
+
+        header.BlockTypes = new string[numBlockTypes];
+        header.BlockTypeIndices = new ushort[numBlocks];
+        header.BlockSizes = new int[numBlocks];
+        blocks = new NifBlock?[numBlocks];
+
+        for (uint i = 0; i < numBlockTypes; i++) {
+            header.BlockTypes[i] = Reader.ReadAdjustedStringLen32();
+        }
+
+        for (uint i = 0; i < numBlocks; i++) {
+            header.BlockTypeIndices[i] = (ushort) (0x7FFFU & Reader.ReadAdjustedUInt16());
```

Comment:
curious if this `&` is actually needed, it seems mostly redundant except truncating the MSB so it would only matter if the value read was > 32767

Reply:
I'm actually not sure on that one. I can try running it without that. It might be because of engine plugin block types now that I think of it.